### PR TITLE
Try to get best usable locale

### PR DESCRIPTION
**Pull request by bcoca**
specially for when you have parameters in unicode but need
  to scrape responses

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

module_utils/locale

### CODE REVIEW

**Comment by mkrizek** [comment]
!needs_collection_redirect

File: lib/ansible/module_utils/common/locale.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c), Ansible Project
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py:1]:
```suggestion
# Copyright: (c) 2021, Ansible Project
```

File: lib/ansible/module_utils/common/locale.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def get_best_parsable_locale(module, preferences=None):
+    '''
+        Attempts to return the best possible locale for parsing output in english
+        useful for scraping output with i18n tools
+
+        :param module: an AnsibleModule instance
+        :param preferences: A list of prefered locales, in order of preference
+        :returns: The first matched prefered locale or 'C' which is the default
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py]:
```suggestion
        Attempts to return the best possible locale for parsing output in English
        useful for scraping output with i18n tools

        :param module: an AnsibleModule instance
        :param preferences: A list of preferred locales, in order of preference
        :returns: The first matched preferred locale or 'C' which is the default
```

File: lib/ansible/module_utils/common/locale.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    '''
+
+    found = 'C'  # default posix, its ascii but always there
+    if preferences is None:
+        # new posix standard or english cause those are messages core team expects
+        # yes, last 2 are same but some systems are weird
+        preferences = ['C.utf8', 'en_US.utf8', 'C', 'POSIX']
+
+    rc, out, err = module.run_command(['locale', '-a'])
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py]: Do you think we should add `get_bin_path` here?

File: lib/ansible/module_utils/common/locale.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+        :param module: an AnsibleModule instance
+        :param preferences: A list of prefered locales, in order of preference
+        :returns: The first matched prefered locale or 'C' which is the default
+    '''
+
+    found = 'C'  # default posix, its ascii but always there
+    if preferences is None:
+        # new posix standard or english cause those are messages core team expects
+        # yes, last 2 are same but some systems are weird
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py]:
```suggestion
        # new POSIX standard or English cause those are messages core team expects
        # yes, the last 2 are the same but some systems are weird
```

File: lib/ansible/module_utils/common/locale.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    rc, out, err = module.run_command(['locale', '-a'])
+    if rc == 0:
+        if out:
+            available = out.strip().splitlines()
+        else:
+            module.warn("No output from locale, defaulting to C, rc=%s: %s" % (rc, to_native(err)))
+    else:
+        module.warn("Unable to get locale information, defaulting to C, rc=%s: %s" % (rc, to_native(err)))
+
+    if available:
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py:44]: if `rc` is not 0 then available will be unbound variable.

File: lib/ansible/module_utils/common/locale.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        # new posix standard or english cause those are messages core team expects
+        # yes, last 2 are same but some systems are weird
+        preferences = ['C.utf8', 'en_US.utf8', 'C', 'POSIX']
+
+    rc, out, err = module.run_command(['locale', '-a'])
+    if rc == 0:
+        if out:
+            available = out.strip().splitlines()
+        else:
+            module.warn("No output from locale, defaulting to C, rc=%s: %s" % (rc, to_native(err)))
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py]: Can we add a parameter like `ignore_errors=silent` or something which matches the other APIs?

File: changelogs/fragments/parseable_locale.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - added new function to module utils to choose best possible locale.
```
Comment (Akasurde) [inline changelogs/fragments/parseable_locale.yml:2]: Would it possible to add test for this change?

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: changelogs/fragments/parseable_locale.yml (hunk 7 above)
Comment (bcoca) [reply changelogs/fragments/parseable_locale.yml:2]: once we agree this is something we want, i'll add unit tests

File: lib/ansible/module_utils/common/locale.py (hunk 6 above)
Comment (bcoca) [reply lib/ansible/module_utils/common/locale.py]: added ignore_warnings

File: lib/ansible/module_utils/common/locale.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/module_utils/common/locale.py]: yes, had already considered it, but had not gotten around to add yet

File: lib/ansible/module_utils/common/locale.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    available = []
+    found = 'C'  # default posix, its ascii but always there
+
+    if preferences is None:
+        # new POSIX standard or English cause those are messages core team expects
+        # yes, the last 2 are the same but some systems are weird
+        preferences = ['C.utf8', 'en_US.utf8', 'C', 'POSIX']
+
+    rc, out, err = module.run_command([locale, '-a'])
```
Comment (Akasurde) [inline lib/ansible/module_utils/common/locale.py:34]: Do you think we should use `get_bin_path` over here?

Comment (bcoca) [reply lib/ansible/module_utils/common/locale.py:34]: that is how `locale` variable gets set, see line 21

Comment (Akasurde) [reply lib/ansible/module_utils/common/locale.py:34]: sorry my bad. I didn't see that.

**Pull request merged**